src/metabase/api/x_ray.clj

```diff
@@ -32,163 +33,133 @@
-       fe/x-ray))
-
-(api/defendpoint GET "/table/:id"
-  "Get x-ray for a `Tield` with ID."
-  [id max_query_cost max_computation_cost]
-  {max_query_cost       MaxQueryCost
-   max_computation_cost MaxComputationCost}
+   max_computation_cost MaxComputationCost
+   model                Model}
   (api/check-403 (costs/enable-xrays))
   (->> id
-       (api/read-check Table)
+       (api/read-check (->model model))
        (fe/extract-features {:max-cost (max-cost max_query_cost
                                                  max_computation_cost)})
        fe/x-ray))
 
-(api/defendpoint GET "/segment/:id"
-  "Get x-ray for a `Segment` with ID."
-  [id max_query_cost max_computation_cost]
-  {max_query_cost       MaxQueryCost
-   max_computation_cost MaxComputationCost}
-  (api/check-403 (costs/enable-xrays))
-  (->> id
-       (api/read-check Segment)
-       (fe/extract-features {:max-cost (max-cost max_query_cost
-                                                 max_computation_cost)})
-       fe/x-ray))
+(defn- virtual-card
+  [query]
+  (->> {:dataset_query query}
+       (merge (card/query->database-and-table-ids query))
+       card/map->CardInstance))
 
-(api/defendpoint GET "/card/:id"
-  "Get x-ray for a `Card` with ID."
-  [id max_query_cost max_computation_cost]
+(api/defendpoint POST "/query"
+  "Get x-ray of a query with no associated model."
+  [max_query_cost max_computation_cost :as {{:keys [database], :as query} :body}]
   {max_query_cost       MaxQueryCost
    max_computation_cost MaxComputationCost}
   (api/check-403 (costs/enable-xrays))
-  (->> id
-       (api/read-check Card)
+  (when-not (= database database/virtual-id)
+    (api/read-check Database database))
+  (->> query
+       virtual-card
        (fe/extract-features {:max-cost (max-cost max_query_cost
                                                  max_computation_cost)})
        fe/x-ray))
 
-(api/defendpoint GET "/compare/fields/:id1/:id2"
-  "Get comparison x-ray for `Field`s with ID1 and ID2."
-  [id1 id2 max_query_cost max_computation_cost]
+(api/defendpoint GET "/compare/:models/:id1/:id2"
+  "Get comparison x-ray for two models of type `:models` with ids :id1 and id2."
+  [models id1 id2 max_query_cost max_computation_cost]
   {max_query_cost       MaxQueryCost
-   max_computation_cost MaxComputationCost}
+   max_computation_cost MaxComputationCost
+   models               Model}
   (api/check-403 (costs/enable-xrays))
   (->> [id1 id2]
-       (map #(api/read-check Field (Integer/parseInt %)))
+       (map #(api/read-check (->model models) (Long/parseLong %)))
        (apply fe/compare-features
               {:max-cost (max-cost max_query_cost max_computation_cost)})
        fe/x-ray))
 
-(api/defendpoint GET "/compare/tables/:id1/:id2"
-  "Get comparison x-ray for `Table`s with ID1 and ID2."
-  [id1 id2 max_query_cost max_computation_cost]
+(api/defendpoint GET "/compare/:model1/:id1/model2/:id2"
```

Comment:
I think this is trying to be a bit too cute. 
Also, shouldn't this be :model2? Why do tests pass if that's the case?

Reply:
I think this is getting too abstract for its own sake. We won't be comparing two arbitrary models with each other, we'll be doing concrete things like "tables w/ segments", "metrics with metrics", etc. I worry that we're going to end up losing sight of what makes each of these specifically useful in an overly generic scaffolding.

Reply:
are there tests for any of these endpoints?